server/src/main/java/io/deephaven/server/plugin/type/JsPluginDistribution.java

```diff
@@ -0,0 +1,137 @@
+
+/**
+ * A {@link JsPlugin} implementation sourced from a distribution directory.
+ */
+public final class JsPluginDistribution extends JsPluginBase {
+
+    private static final String PACKAGE_JSON = "package.json";
+
+    /**
+     * Creates a new js plugin distribution. Assumes that {@value PACKAGE_JSON} exists in {@code distributionDir}. The
+     * {@link #name() name}, {@link #version() version}, and {@link #main() main} from {@value PACKAGE_JSON} will be
+     * used.
+     *
+     * @param distributionDir the distribution directory
+     * @return the js plugin distribution
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPluginDistribution fromDistribution(Path distributionDir) throws IOException {
+        final Path packageJson = distributionDir.resolve(PACKAGE_JSON);
+        try (final InputStream in = new BufferedInputStream(Files.newInputStream(packageJson))) {
+            final PackageJson p = new ObjectMapper()
+                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                    .readValue(in, PackageJson.class);
+            // Note: we could provide an option, or do it by default, to only copy over p.main, or p.main.parent, if we
+            // wanted to minimize the amount of data we re-expose.
+            return new JsPluginDistribution(distributionDir, p.name, p.version, p.main);
+        }
+    }
+
+    private final Path distributionDir;
+    private final String name;
+    private final String version;
+    private final String main;
+
+    /**
+     * Creates a new js plugin distribution.
+     *
+     * <p>
+     * Note: unlike {@link #fromDistribution(Path)}, {@code distributionDir} does not need to contain
+     * {@value PACKAGE_JSON}.
+     *
+     * @param distributionDir the distribution directory
+     * @param name the name, see {@link JsPlugin#name()}
+     * @param version the version, see {@link JsPlugin#version()}
+     * @param main the main, see {@link JsPlugin#main()}
+     */
+    public JsPluginDistribution(Path distributionDir, String name, String version, String main) {
+        this.distributionDir = Objects.requireNonNull(distributionDir);
+        this.name = Objects.requireNonNull(name);
+        this.version = Objects.requireNonNull(version);
+        this.main = Objects.requireNonNull(main);
+    }
+
+    @Override
+    public String name() {
+        return name;
+    }
+
+    @Override
+    public String version() {
+        return version;
+    }
+
+    @Override
+    public String main() {
+        return main;
+    }
+
+    @Override
+    public void copyTo(Path destination) throws IOException {
+        copyRecursive(distributionDir, destination);
+    }
+
+    private static void copyRecursive(Path src, Path dst) throws IOException {
+        Files.createDirectories(dst.getParent());
+        Files.walkFileTree(src, new CopyRecursiveVisitor(src, dst));
+    }
+
+    static final class PackageJson {
```

Comment:
conventions should be documented, either in this repo or possibly in https://github.com/deephaven/js-plugin-template? esp since we hardcode the name, and fail on unknown properties (so "real" `package.json` files will fail it)

Reply:
There is a bit of docs in `io.deephaven.server.plugin.type.JsPluginDistribution#fromPackageJsonDistribution` saying that name, version, and main are used from package.json. I think linking out to this repository makes sense though, will do that.

Reply:
Sure, but that's in the guts of the java code - the author of a js plugin is going to look at js-plugin docs, not javadocs of a given class on the platform?

Reply:
I'll add a link at `JsPlugin` level.

Reply:
What I'm saying is that we need more than Javadoc for this - a JS/TS developer shouldn't have to already know how java/py works to be able to figure this out.

Reply:
> I don't expect JS/TS dev being able to necessarily decipher through this either. I hope the external js plugin documentation will be well suited for them, and then we can have docs for "here's how to convert it to PyPi package or jar for easy inclusion".

To confirm, we don't expect that a JS dev will be able to work out how to produce a jar/zip or a wheel, or what the format should be of the `package.json`? I don't think that's what you're saying, but it seems to be.

I'm not asking for "lots of java tutorials", but just "here's the shortest path-to-success to produce a js plugin for deephaven, what files matter, what fields matter, why your own package.json isn't valid, just a subset of it". Mostly "you can have a name, version, and main, we'll effectively ignore the version, and use the main as a relative path from your package.json file", etc. Such a plugin might have no py and no java, and no plan to add either, but wants to know the degrees of freedom they should have? 

Maybe you're saying you'll file something against js-plugins to write this up? That seems like a good place, as long as it is agnostic to the language (since this should be strictly declarative)?